Add message checks for NetworkConnectionToWebProcess::deleteCookie

pratiksha-99 opened the pull request:
#### 55cce6403f814f7334bb36ff8f14d7bb6323a55f
<pre>
Add message checks for NetworkConnectionToWebProcess::deleteCookie
<a href="https://bugs.webkit.org/show_bug.cgi?id=291659">https://bugs.webkit.org/show_bug.cgi?id=291659</a>
<a href="https://rdar.apple.com/149288315">rdar://149288315</a>

Reviewed by Sihui Liu.

A valid web content process should not send invalid arguments.

* LayoutTests/ipc/coreipc.js:
(export.resolveAlias):
(export.ArgumentSerializer):
* LayoutTests/ipc/empty-cookie-invalid-test-crash-expected.txt: Added.
* LayoutTests/ipc/empty-cookie-invalid-test-crash.html: Added.
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::deleteCookie):

Canonical link: <a href="https://commits.webkit.org/293847@main">https://commits.webkit.org/293847@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4eaf3fa913d0290325700ae650343c437807979f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100028 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19676 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9968 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105156 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50609 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19982 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28149 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76148 "Found 1 new test failure: imported/w3c/web-platform-tests/service-workers/service-worker/redirected-response.https.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33224 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103035 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15259 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90336 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56508 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15069 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8331 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49978 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84988 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8416 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107516 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27141 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19880 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85101 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27504 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86541 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84628 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29303 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7034 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/20977 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16286 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27078 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32307 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26889 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30205 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28448 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on LayoutTests/ipc/coreipc.js:
```diff
@@ -915,14 +970,61 @@ export class ArgumentParser {
         }
         return [position, values];
     }
+    /*
+    static serializeStdArray(innerType, argument) {
+        const splitType = ArgumentSerializer.splitTemplateType(innerType);
+        if(splitType.length != 2) {
+            throw new SerializationError('std::array does not have a fixed cardinality')
+        }
+        const size = Number(splitType[1]);
+        innerType = splitType[0];
+        if(Array.isArray(argument) || argument.length != size) {
+            const result = [];
+            const argumentDefinition = {
+                type: innerType,
+            };
+            let count = 0;
+            for(const element of argument) {
+                argumentDefinition.name = `element#${ count }`;
+                count += 1;
+                result.push([ ArgumentSerializer.serializeArgument(argumentDefinition, element) ]);
+            }
+            return result;
+        }
+        throw new SerializationError('argument of std::pair type is not an array');
+    }
+    */
```
Delete this?

pratiksha-99 replied:
Not my code, it came from script made by SEAR for fuzzing. We dont maintain it.

szewai commented on LayoutTests/ipc/empty-cookie-invalid-test-crash.html:
```diff
@@ -0,0 +1,18 @@
+<!-- webkit-test-runner [ IPCTestingAPIEnabled=true ] -->
+<p>This test passes if WebKit does not crash.</p>
+<script>
+   if (window.testRunner) {
```
Wrong indentation

szewai commented on LayoutTests/ipc/coreipc.js:
```diff
@@ -453,7 +453,7 @@ export class ArgumentSerializer {
         const splitType = ArgumentSerializer.splitTemplateType(innerType);
         if(splitType.length != 2) {
-            throw new SerializationError('std::pair does not have a fixed cardinality')
+            throw new SerializationError('std::array does not have a fixed cardinality')
```
Maybe logging the actual length in error?

pratiksha-99 replied:
same, not my code.

szewai commented on LayoutTests/ipc/coreipc.js:
```diff
@@ -491,32 +491,56 @@ export class ArgumentSerializer {
+            return ArgumentSerializer.serializePair(innerType, argument);
+        }
+
+        if(argument['key'] !== undefined && argument['value'] !== undefined) {
```
I think we normally early return on error, like:
```
if (argument['key'] === undefined || argument['value'] === undefined)
    throw new SerializationError(...)

const splitTemplateType = ArgumentSerializer.splitTemplateType(innerType);
...
```

(And I think there should be space between `if` and `(`, but it seems all conditions are missing that)

szewai commented on LayoutTests/ipc/coreipc.js:
```diff
@@ -491,32 +491,56 @@ export class ArgumentSerializer {
+        let variantType = undefined;
+        let variantIndex = undefined;
+
+        if(Object.hasOwn(argument, 'variantType')) {
+            variantType = argument.variantType;
+            variantIndex = variantTypes.indexOf(argument.variantType);
+
+            if(variantIndex == -1) {
+                throw new SerializationError(`type '${ variantType }' is not valid for variant (${ variantTypes })`);
+            }
+        }
+
+        if(Object.hasOwn(argument, 'variantIndex')) {
+            variantIndex = argument.variantIndex;
+            variantType = variantTypes[variantIndex];
```
Do we want to make sure `argument` does not have both? 
Or if it can have both, maybe we should verify the  computed `argument.variantIndex` matches `variantIndex` instead of just overwriting `variantIndex`?

szewai reviewed (approved):
LGTM with nits

pratiksha-99 committed:
Add message checks for NetworkConnectionToWebProcess::deleteCookie
https://bugs.webkit.org/show_bug.cgi?id=291659
rdar://149288315

Reviewed by Sihui Liu.

A valid web content process should not send invalid arguments.

* LayoutTests/ipc/coreipc.js:
(export.resolveAlias):
(export.ArgumentSerializer):
* LayoutTests/ipc/empty-cookie-invalid-test-crash-expected.txt: Added.
* LayoutTests/ipc/empty-cookie-invalid-test-crash.html: Added.
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::deleteCookie):

Canonical link: https://commits.webkit.org/293847@main
